File: kuber/kuber/dockerclient/process.py
import asyncio
import traceback

from docker.models.containers import ExecResult
from dockerclient import container


def execute_command(name: str, cmd: str, dir: str) -> ExecResult:
    try:
        cont = container.get_container(name=name)
        output = cont.exec_run(cmd, workdir=dir, stream=True)
    except Exception:
        pass
    return output


class Process:
    def __init__(self, container_name, command, dir) -> None:
        self.container_name = container_name
        self.command = command
        self.dir = dir
        self.proc = None
        self.pid = None

    async def start(self):
        try:
            self.proc = await asyncio.subprocess.create_subprocess_shell(
                f"docker exec -i {self.container_name} {self.command}",
                cwd=self.dir,
                stdin=asyncio.subprocess.PIPE,
                stdout=asyncio.subprocess.PIPE,
            )
        except Exception as e:
            print(str(e))

    async def get_pid(self):
        tproc = await asyncio.subprocess.create_subprocess_shell(
            f"docker exec -i {self.container_name} pgrep -a .",
            stdout=asyncio.subprocess.PIPE,
        )
        stdout, _ = await tproc.communicate()
        lines = stdout.decode().split("\n")
        for line in lines:
            if line != "":
                details = line.split()
                pid = details[0]
                cmd = " ".join(details[1:])
                if cmd == self.command:
                    self.pid = pid
                    break

    async def is_alive(self) -> bool:
        if not self.pid:
            return False
        tproc = await asyncio.subprocess.create_subprocess_shell(
            f"docker exec -i {self.container_name} ps -q {self.pid} -o state --no-headers",
            stdout=asyncio.subprocess.PIPE,
        )
        stdout, err = await tproc.communicate()
        lines = stdout.decode().split("\n")
        print(f"lines - {lines}", err)
        if err:
            return False
        return True

    async def readline(self) -> str:
        try:
            line = await asyncio.wait_for(self.proc.stdout.readline(), timeout=1)
        except TimeoutError:
            return ""
        return line

    async def proc_kill(self) -> bool:
        tproc = await asyncio.subprocess.create_subprocess_shell(
            f"docker exec -i {self.container_name} kill {self.pid}",
            stdout=asyncio.subprocess.PIPE,
        )
        out = await tproc.wait()
        if out == 0:
            return True
        return False

    def write_input(self, input: str) -> bool:
        try:
            self.proc.stdin.write(f"{input}\n".encode("utf-8"))
            self.proc.stdin.write(b"\n")
            return True
        except Exception:
            return False

    def kill(self):
        try:
            self.proc.kill()
        except Exception:
            pass

    async def get_full_output(self):
        stdout, err = await self.proc.communicate()
        if not err:
            return stdout
        return err

File: executor/docker/process.go
package docker

import (
	"bufio"
	"context"
	"errors"
	"io"
	"os/exec"
	"strings"
	"sync"
	"time"
)

type Process struct {
	ContainerName string
	Command       []string
	Dir           string
	Proc          *exec.Cmd
	Pid           string
	Lock          sync.Mutex
	stdin         io.WriteCloser
	stdout        io.Reader
}

func NewProcess(containerName string, command []string, dir string) *Process {
	return &Process{
		ContainerName: containerName,
		Command:       command,
		Dir:           dir,
	}
}

func (p *Process) Start() error {
	p.Lock.Lock()
	defer p.Lock.Unlock()

	cmdPrefix := []string{"exec", "-i", p.ContainerName}
	cmd := append(cmdPrefix, p.Command...)
	p.Proc = exec.Command("docker", cmd...)
	p.Proc.Dir = p.Dir

	var err error
	p.stdin, err = p.Proc.StdinPipe()
	if err != nil {
		return err
	}
	p.stdout, err = p.Proc.StdoutPipe()
	if err != nil {
		return err
	}

	return p.Proc.Start()
}

func (p *Process) GetPID() (string, error) {
	p.Lock.Lock()
	defer p.Lock.Unlock()

	cmd := exec.Command("docker", "exec", "-i", p.ContainerName, "pgrep", "-a", ".")
	output, err := cmd.Output()
	if err != nil {
		return "", err
	}

	lines := strings.Split(strings.TrimSpace(string(output)), "\n")
	for _, line := range lines {
		parts := strings.Fields(line)
		if len(parts) >= 2 {
			pid := parts[0]
			cmd := strings.Join(parts[1:], " ")
			if cmd == strings.Join(p.Command, " ") {
				p.Pid = pid
				return pid, nil
			}
		}
	}

	return "", nil
}

func (p *Process) IsAlive() bool {
	p.Lock.Lock()
	defer p.Lock.Unlock()

	if p.Pid == "" {
		return false
	}

	cmd := exec.Command("docker", "exec", "-i", p.ContainerName, "ps", "-p", p.Pid)
	err := cmd.Run()
	return err == nil
}

func (p *Process) ReadLine() (string, error) {
	p.Lock.Lock()
	defer p.Lock.Unlock()

	reader := bufio.NewReader(p.stdout)
	line, _, err := reader.ReadLine()
	return string(line), err
}

func (p *Process) ReadLineWithTimeout(timeout time.Duration) (string, error) {
	p.Lock.Lock()
	defer p.Lock.Unlock()

	ctx, cancel := context.WithTimeout(context.Background(), timeout)
	defer cancel()

	reader := bufio.NewReader(p.stdout)
	var line string
	var err error

	go func() {
		select {
		case <-ctx.Done():
			err = errors.New("Timeout")
		default:
			var lineBytes []byte
			lineBytes, _, err = reader.ReadLine()
			line = string(lineBytes)
		}
	}()

	<-ctx.Done()
	return line, err
}

func (p *Process) ReadOutput() (string, error) {
	p.Lock.Lock()
	defer p.Lock.Unlock()

	output, err := io.ReadAll(p.stdout)
	if err != nil {
		return "", err
	}

	return string(output), nil
}

func (p *Process) Kill() error {
	p.Lock.Lock()
	defer p.Lock.Unlock()

	if p.Proc.Process != nil {
		return p.Proc.Process.Kill()
	}
	return nil
}

func (p *Process) WriteInput(input string) error {
	p.Lock.Lock()
	defer p.Lock.Unlock()

	_, err := io.WriteString(p.stdin, input+"\n")
	if err != nil {
		return err
	}
	return nil
}
